[lang] start to add support for arithmetic expressions

tstack opened the pull request:
This change contains a start at adding support for arithmetic
expressions in the query language (e.g. a + b).  The currently
supported operators are add, subtract, multiply, and divide
for numeric operands.  There is also a field query operator
(i.e. <expr> as <name>).

TODO - Update the README

Future work:
  - Support function calls
  - Floating point literals

Files:
  * Cargo.toml, Cargo.lock: Bump ordered-float version
  * data.rs: Add arithmetic ops for Value
  * errors.rs: Add a MissingOperand error
  * lang.rs: Add parser for arithmetic operations
  * operator.rs: Remove EvaluatableBorrowed since it
    doesn't work with intermediate values.
  * typecheck.rs: Convert arithmetic ops.

tstack commented:
I wasn't planning anything else for this PR, let me know if you want anything else done.

rcoh commented on src/data.rs:
```diff
@@ -524,4 +605,63 @@ mod tests {
+    fn arithmetic() {
+        assert_eq!(
+            Value::Float(OrderedFloat(5.)),
+            (Value::Float(OrderedFloat(4.)) + Value::Float(OrderedFloat(1.))).unwrap()
```
if you like this (and a number of other tests can be cleaned up with the helper `Value::from_float`)

rcoh commented on src/typecheck.rs:
```diff
@@ -75,6 +86,15 @@ impl TypeCheck<operator::Expr> for lang::Expr {
+                    > {
+                        left: Box::new((*left).type_check(error_builder)?),
+                        right: Box::new((*right).type_check(error_builder)?),
+                        operator: arith_op.into(),
```
if it's not too inconvenient, I have a mild preference for implementing `type_check` instead of `from`. I used to have `into()s` all over the place and it got super confusing & had inference problems so I switched everything to `type_check`.

tstack commented:
> Especially impressed by the parser -- I had tried fruitlessly a while ago to get nom to work on a recursive grammar and could never quite get it working.

I cribbed it from the nom example and some other expression parsers I have worked on.  It works for now, but will probably not work when support for functions are added.  The `peek!(did_you_mean)` stuff for operators causes plain expressions to be rejected.  Something will need to be figured out there.

I've pushed one more commit that I think addresses your suggestions.

rcoh commented:
Yeah... I'm considering pulling that into typecheck by emitting an "unknown
operator"

I used nom5 in another project and it's a bit nicer (no more macros). I
will probably pull lang into its own (internal) crate soon to facilitate
that rewrite

On Fri, Jul 31, 2020, 1:02 AM Tim Stack <notifications@github.com> wrote:

> Especially impressed by the parser -- I had tried fruitlessly a while ago
> to get nom to work on a recursive grammar and could never quite get it
> working.
>
> I cribbed it from the nom example and some other expression parsers I have
> worked on. It works for now, but will probably not work when support for
> functions are added. The peek!(did_you_mean) stuff for operators causes
> plain expressions to be rejected. Something will need to be figured out
> there.
>
> I've pushed one more commit that I think addresses your suggestions.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rcoh/angle-grinder/pull/110#issuecomment-666922194>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADYKZ3LJKW2RZHDYCAWHG3R6JF6BANCNFSM4PHJM5KQ>
> .
>
